src/compiler/tsbuild.ts

```diff
@@ -1182,6 +1183,97 @@ namespace ts {
 
         function buildAllProjects(): ExitStatus {
             if (options.watch) { reportWatchStatus(Diagnostics.Starting_compilation_in_watch_mode); }
+            const originalReadFile = host.readFile;
+            const originalFileExists = host.fileExists;
+            const originalDirectoryExists = host.directoryExists;
+            const originalCreateDirectory = host.createDirectory;
+            const originalWriteFile = host.writeFile;
+            const originalGetSourceFile = host.getSourceFile;
+            const readFileCache = createMap<string | false>();
+            const fileExistsCache = createMap<boolean>();
+            const directoryExistsCache = createMap<boolean>();
+            const sourceFileCache = createMap<SourceFile>();
+            const savedReadFileWithCache = readFileWithCache;
+
+            // TODO:: In watch mode as well to use caches for incremental build once we can invalidate caches correctly and have right api
+            // Override readFile for json files and output .d.ts to cache the text
+            readFileWithCache = fileName => {
+                const key = toPath(fileName);
+                const value = readFileCache.get(key);
+                if (value !== undefined) return value || undefined;
+                return setReadFileCache(key, fileName);
+            };
+            const setReadFileCache = (key: Path, fileName: string) => {
+                const newValue = originalReadFile.call(host, fileName);
+                readFileCache.set(key, newValue || false);

```

Comment:
doesn't this imply an entry in fileExistsCache since the keys are the same?

Reply:
host.readFile can act differently (failing only on readFile by sending undefined so have avoided changing file exists because of readFile)